### PR TITLE
Setup base runner

### PR DESCRIPTION
**Pull request by nully0x**
This pull request introduces significant enhancements to our test runner service, improving its functionality, flexibility, and output. Key changes include:

1. **Rust Language Support**
   - Added capability to detect and run Typescript and Rust challenge entrypoint
   - Implemented a lightweight Rust Docker image for testing Typescript and Rust code

2. **Docker-based Test Isolation**
   - Replaced Nix with Docker for running tests, enhancing isolation and reproducibility
   - Implemented Docker image building and container management within the test runner

3. **Event Type Differentiation**
   - Implemented distinct event types for test triggers and test results
   - Added 'test_result' as a specific event type for reporting test outcomes

4. **Optimized Docker Images**
   - Utilized Alpine-based images to reduce overall container size and improve build times

5. **Enhanced Error Handling and Clean-up**
   - Implemented robust error catching and reporting throughout the test process
   - Added clean-up procedures for Docker resources and cloned repositories

6. **Configuration Improvements**
   - Updated configuration to support multiple languages and their specific Docker setups

One improvement to be implemented is caching users code session(not clearing the docker  images/container immediately the code executes), as  pulling depedencies every time a user push their code can use quite a lot of bandwith and take some time in between. (though users can run their code locally before pushing)

### CODE REVIEW

File: src/config/config.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  dockerfilePath: string;
+  runCommand: string;
+}
+
+export const LANGUAGE_CONFIGS: Record<string, LanguageConfig> = {
+  typescript: {
+    language: "typescript",
+    dockerfilePath: path.join(
+      __dirname,
+      "../../supportedLangaugeDockerfiles/typescript/Dockerfile",
```
Comment (daviroo) [inline src/config/config.ts]:
```suggestion
      "../../supportedLanguageDockerfiles/typescript/Dockerfile",
```

File: src/config/config.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    ),
+    runCommand: "/app/.hxckr/run.sh",
+  },
+  rust: {
+    language: "rust",
+    dockerfilePath: path.join(
+      __dirname,
+      "../../supportedLangaugeDockerfiles/rust/Dockerfile",
```
Comment (daviroo) [inline src/config/config.ts]:
```suggestion
      "../../supportedLanguageDockerfiles/rust/Dockerfile",
```

File: supportedLangaugeDockerfiles/rust/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+# Copy the entire application
+COPY . .
+
+# Build dependencies to cache them
+RUN cargo build --release
+
+# Make the run.sh script executable
+RUN chmod +x /app/.hxckr/run.sh
+
+# Make the run.sh script executable
```
Comment (daviroo) [inline supportedLangaugeDockerfiles/rust/Dockerfile]: I guess this comment is incorrect?  Or is the command below incorrect?

**Review by daviroo — CHANGES_REQUESTED** [review]
Just some small comments @nully0x - aside from that it looks good to go.

File: supportedLangaugeDockerfiles/rust/Dockerfile (hunk 3 above)
Comment (nully0x) [reply supportedLangaugeDockerfiles/rust/Dockerfile]: Yes it's correct, gives the container a bash entry to be able to run this `runCommand: "/app/.hxckr/run.sh"` which is in the dir all challenge users will be taking on.

File: src/config/config.ts (hunk 1 above)
Comment (nully0x) [reply src/config/config.ts]: @daviroo can't pinpoint what changed here, can you elaborate?

File: src/config/config.ts (hunk 2 above)
Comment (nully0x) [reply src/config/config.ts]: @daviroo can pinpoint what changed here, can you elaborate?

Comment (daviroo) [reply src/config/config.ts]: Language is misspelled.

File: src/config/config.ts (hunk 1 above)
Comment (daviroo) [reply src/config/config.ts]: Language is misspelled.

Comment (nully0x) [reply src/config/config.ts]: Damn, I got a bug here cause of this, I was why the pushed code was not compiling, ha ha, I copied the name in the folder itself and paste it here. Didn't notice it was misspelled. Thanks for the catch, will change the folder name too.

**Review by daviroo — APPROVED** [review]

**Pull request merged**